File: programmers/new/level2/스킬트리.java
class Solution {
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        int[] skillIdx = new int[128];
        for(int i = 'A'; i <= 'Z'; ++i){
            skillIdx[i] = -1;
        }
        for(int i = 0; i < skill.length(); ++i){
            char c = skill.charAt(i);
            skillIdx[c] = i;
        }
        f: for(int i = 0; i < skill_trees.length; ++i){
            String sk = skill_trees[i];
            int lastIdx = 0;
            for(int j = 0; j < sk.length(); ++j){
                if(skillIdx[sk.charAt(j)] != -1){
                    if(skillIdx[sk.charAt(j)] == lastIdx){
                        lastIdx++;
                    }else{
                        continue f;
                    }
                }
            }
            answer++;
        }
        return answer;
    }
}
public class 스킬트리 {
	public static void main(String[] args) throws Exception{
		
	}
}

File: programmers/js/new/level2/스킬트리.js
function solution(skill, skill_trees) {
    var answer = 0;
    const set = new Set();
    skill.split("").forEach((c) => {
        set.add(c);
    });
    skill_trees.forEach((st) => {
        if (check(set, skill, st)) {
            answer++;
        }
    });
    return answer;
}

const check = (set, skill, st) => {
    let skillIdx = 0;
    const arr = st.split("");
    for (let i = 0; i < arr.length; ++i) {
        const c = arr[i];
        if (set.has(c)) {
            if (skill.charAt(skillIdx) === c) {
                skillIdx++;
            } else {
                return false;
            }
        }
    }
    return true;
};
